app/server/lib/OIDCConfig.ts

```diff
@@ -52,26 +63,150 @@
+} from 'openid-client';
 import { Sessions } from './Sessions';
 import log from 'app/server/lib/log';
-import { appSettings } from './AppSettings';
+import { AppSettings, appSettings } from './AppSettings';
 import { RequestWithLogin } from './Authorizer';
 import { UserProfile } from 'app/common/LoginSessionAPI';
+import { SendAppPage } from './sendAppPage';
+import { StringUnion } from 'app/common/StringUnion';
 
 const CALLBACK_URL = '/oauth2/callback';
 
+const EnabledProtection = StringUnion(
+  "STATE",
+  "NONCE",
+  "PKCE",
+);
+EnabledProtection.setErrMessageBuilder(
+  (actual, _, expectedAsArray: string[]) =>
+    `OIDC: Invalid protection in GRIST_OIDC_IDP_ENABLED_PROTECTIONS: ${actual}.`+
+    ` Expected at least one of these values: "${expectedAsArray.join(",")}"`
+);
+
+type EnabledProtectionString = typeof EnabledProtection.type;
+type ProtectionPropertyName = 'nonce' | 'state' | 'code_verifier';
+type Protections = Partial<Record<ProtectionPropertyName, string>>;
+
+class SingleProtection {
+  private static _generationMethodsByPropertyName = {
+    state: () => generators.state(),
+    code_verifier: () => generators.codeVerifier(),
+    nonce: () => generators.nonce(),
+  };
+
+  constructor(
+    public readonly propertyName: ProtectionPropertyName,
+    public readonly errMsgIfOmitted: string,
+  ) {}
+
+  public generate() {
+    const generationMethod = SingleProtection._generationMethodsByPropertyName[this.propertyName];
+    return generationMethod();
+  }
+
+  public assertExistsInSession(sessionInfo: Record<string, string>) {
+    if (!sessionInfo[this.propertyName]) {
+      throw new Error(this.errMsgIfOmitted);
+    }
+  }
+}
+
+class ProtectionsManager {
+  private _protections: SingleProtection[] = [];
+
+  constructor(private _enabledProtections: Set<EnabledProtectionString>) {
+    if (this._enabledProtections.has('STATE')) {
+      this._protections.push(new SingleProtection('state', 'Login or logout failed to complete'));
+    }
+    if (this._enabledProtections.has('NONCE')) {
+      this._protections.push(new SingleProtection('nonce', 'Login is stale'));
+    }
+    if (this._enabledProtections.has('PKCE')) {
+      this._protections.push(new SingleProtection('code_verifier', 'Login is stale'));
+    }
+  }
+
+  public generate(): Protections {
+    const protections: Protections = {};
+    for (const singleProtection of this._protections) {
+      protections[singleProtection.propertyName] = singleProtection.generate();
+    }
+    return protections;
+  }
+
+  public forgeProtectionParamsForAuthUrl(protections: Protections) {
+    return {
+      state: protections.state,
+      nonce: protections.nonce,
+      code_challenge: protections.code_verifier ?
+        generators.codeChallenge(protections.code_verifier) :
+        undefined,
+      code_challenge_method: protections.code_verifier ? 'S256' : undefined,
+    };
+  }
+
+  public getCallbackChecksFromSessionInfo(sessionInfo: Record<string, any>) {
```

Comment:
Better to use a stricter type here, either replace `any` by `string` or maybe OK to use `SessionObj['oidc']` (as in my gist)?